### PR TITLE
LPS-170221 add validation for duplicate field references

### PR DESCRIPTION
**Pull request by devpedrotavares**
[LPS-170221](https://issues.liferay.com/browse/LPS-170221) The bug was happening because there was no validation for if the fields of the dataDefinition had duplicate references. 
The solution was to create a new method to do this validation and have it throw a custom exception.

### CODE REVIEW

**Comment by devpedrotavares** [comment]
ci:test:sf

**Comment by devpedrotavares** [comment]
ci:test:relevant

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java [hunk 1]
```diff
@@ -322,6 +324,33 @@ private void _validateDDMFormFieldPropertyValue(
...
+		Set<String> duplicateDDMFormFieldReferences = new HashSet<>();
+
+		for (DDMFormField ddmFormField : ddmFormFields) {
+			if (!ddmFormFieldReferences.add(ddmFormField.getFieldReference())) {
+				duplicateDDMFormFieldReferences.add(
+					ddmFormField.getFieldReference());
+			}
+
+			_validateDDMFormFieldReferences(
+				ddmFormField.getNestedDDMFormFields());
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java:345]: @devpedrotavares since you are not using the same `duplicateDDMFormFieldReferences` when validating the `nestedDDMFormFields`, the following scenario will not work:

[Gravação de tela de 09-12-2022 09:21:58.webm](https://user-images.githubusercontent.com/34186632/206701747-8e04dbdc-09f8-4c1e-b5b4-88053af5c276.webm)

Comment (carolmariaabb) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java:345]: I would suggest something like:

```
_validateDDMFormFieldReferences(
	ddmFormFields, new HashSet<>(), new HashSet<>());

...

private void _validateDDMFormFieldReferences(
		List<DDMFormField> ddmFormFields,
		Set<String> ddmFormFieldReferences,
		Set<String> duplicateDDMFormFieldReferences)
	throws DDMFormValidationException {

	for (DDMFormField ddmFormField : ddmFormFields) {
		if (!ddmFormFieldReferences.add(ddmFormField.getFieldReference())) {
			duplicateDDMFormFieldReferences.add(
				ddmFormField.getFieldReference());
		}

		_validateDDMFormFieldReferences(
			ddmFormField.getNestedDDMFormFields(), ddmFormFieldReferences,
			duplicateDDMFormFieldReferences);
	}

	if (SetUtil.isNotEmpty(duplicateDDMFormFieldReferences)) {
		throw new MustNotDuplicateFieldReference(
			duplicateDDMFormFieldReferences);
	}
}
```

Could you take a look at this?

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java [hunk 2]
```diff
@@ -322,6 +324,33 @@ private void _validateDDMFormFieldPropertyValue(
 		}
 	}
 
+	private void _validateDDMFormFieldReferences(
+			List<DDMFormField> ddmFormFields)
+		throws DDMFormValidationException {
+
+		if (ddmFormFields.isEmpty()) {
+			return;
+		}
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java:333]: @devpedrotavares we are already checking it [here](https://github.com/liferay-forms/liferay-portal/pull/2052/files#diff-c312a5bafbece2a3f3bf36526bca133aceb5ab11d1288f52d143134ce74182ffR89-R90)

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java (hunk 1 above)
Comment (devpedrotavares) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java]: That's right! I did something similar. But I created a helper method, so that all the duplicated references caught will be added to the set.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java (hunk 2 above)
Comment (devpedrotavares) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java]: Okay! I removed the unnecessary check.

**Comment by devpedrotavares** [comment]
ci:test:sf

File: modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/jaxrs/exception/mapper/DataDefinitionMustNotDuplicateFieldReferenceValidationExceptionMapper.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+		DataDefinitionValidationException.MustNotDuplicateFieldReference
+			mustNotDuplicateFieldReference) {
+
+		Set<String> duplicatedFieldReferences =
+			mustNotDuplicateFieldReference.getDuplicatedFieldReferences();
+
+		Stream<String> stream = duplicatedFieldReferences.stream();
+
+		return new Problem(
+			stream.collect(Collectors.joining(StringPool.COMMA)),
```
Comment (carolmariaabb) [inline modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/jaxrs/exception/mapper/DataDefinitionMustNotDuplicateFieldReferenceValidationExceptionMapper.java:58]: @devpedrotavares is it possible to not use `stream` here (seee this [slack message](https://liferay.slack.com/archives/CL69D8RUG/p1649806046880829) for more details)?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java [hunk 4]
```diff
@@ -172,6 +173,29 @@ public void testDuplicateFieldName() throws Exception {
...
 
+	@Test(expected = MustNotDuplicateFieldReference.class)
+	public void testDuplicateFieldReference() throws Exception {
+		DDMForm ddmForm = DDMFormTestUtil.createDDMForm(
+			createAvailableLocales(LocaleUtil.US), LocaleUtil.US);
+
+		DDMFormField ddmFormField1 = _createDDMFormField(
+			"Name1", DDMFormFieldType.TEXT, "Reference1");
+
+		ddmForm.addDDMFormField(ddmFormField1);
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java:184]: @devpedrotavares please inline, i.e.:
```
ddmForm.addDDMFormField(_createDDMFormField("Name1", DDMFormFieldType.TEXT, "Reference1"));
```

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java [hunk 5]
```diff
@@ -172,6 +173,29 @@ public void testDuplicateFieldName() throws Exception {
...
+
+		DDMFormField ddmFormField2 = _createDDMFormField(
+			"Name2", DDMFormFieldType.TEXT, "Reference2");
+
+		DDMFormField ddmFormField3 = _createDDMFormField(
+			"Name3", DDMFormFieldType.TEXT, "Reference1");
+
+		ddmFormField2.addNestedDDMFormField(ddmFormField3);
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java:192]: @devpedrotavares please inline, i.e.:
```
ddmFormField.addNestedDDMFormField(_createDDMFormField("Name3", DDMFormFieldType.TEXT, "Reference1"));
```

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java [hunk 6]
```diff
@@ -569,6 +593,16 @@ protected Set<Locale> createAvailableLocales(Locale... locales) {
 		return DDMFormTestUtil.createAvailableLocales(locales);
 	}
 
+	private DDMFormField _createDDMFormField(
+		String name, String type, String reference) {
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java:597]: @devpedrotavares please sort this alphabetically

Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java:597]: @devpedrotavares please rename `reference` to `fieldReference`

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java [hunk 7]
```diff
@@ -160,13 +160,13 @@ public void testDuplicateFieldName() throws Exception {
 			createAvailableLocales(LocaleUtil.US), LocaleUtil.US);
 
 		ddmForm.addDDMFormField(
-			new DDMFormField("Name1", DDMFormFieldType.TEXT));
+			_createDDMFormField("Name1", DDMFormFieldType.TEXT, "Reference1"));
 
-		DDMFormField name2DDMFormField = new DDMFormField(
-			"Name2", DDMFormFieldType.TEXT);
+		DDMFormField name2DDMFormField = _createDDMFormField(
+			"Name2", DDMFormFieldType.TEXT, "Reference2");
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java:166]: @devpedrotavares can you rename `name2DDMFormField` to `ddmFormField`?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java [hunk 8]
```diff
@@ -149,6 +151,24 @@ protected void validateDDMFormFieldValidationExpression(
 		}
 	}
 
+	private void _findDuplicatedDDMFormFieldReferences(
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java:154]: @devpedrotavares please rename it to `_addDuplicatedDDMFormFieldReferences`

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java [hunk 9]
```diff
@@ -322,6 +342,21 @@ private void _validateDDMFormFieldPropertyValue(
...
+
+		Set<String> duplicatedDDMFormFieldReferences = new HashSet<>();
+
+		_findDuplicatedDDMFormFieldReferences(
+			ddmFormFields, new HashSet<>(), duplicatedDDMFormFieldReferences);
+
+		if (SetUtil.isNotEmpty(duplicatedDDMFormFieldReferences)) {
+			throw new MustNotDuplicateFieldReference(
+				duplicatedDDMFormFieldReferences);
+		}
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java:357]: @devpedrotavares since we're only calling this once, we can move this code to the `validate` method (right after the `_validateDDMFormFieldNames` call).

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java [hunk 10]
```diff
@@ -149,6 +151,24 @@ protected void validateDDMFormFieldValidationExpression(
 		}
 	}
 
+	private void _findDuplicatedDDMFormFieldReferences(
+		List<DDMFormField> ddmFormFields, Set<String> ddmFormFieldReferences,
+		Set<String> duplicatedDDMFormFieldReferences) {
+
+		for (DDMFormField ddmFormField : ddmFormFields) {
+			if (!ddmFormFieldReferences.add(
+					StringUtil.toLowerCase(ddmFormField.getFieldReference()))) {
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java:160]: @devpedrotavares why did you add the `StringUtil.toLowerCase`? Shouldn't it be case-sensitive?  Could you leave the same behavior as when we validate the name?

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

File: modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/jaxrs/exception/mapper/DataDefinitionMustNotDuplicateFieldReferenceValidationExceptionMapper.java (hunk 3 above)
Comment (devpedrotavares) [reply modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/jaxrs/exception/mapper/DataDefinitionMustNotDuplicateFieldReferenceValidationExceptionMapper.java]: Yes it is. I updated it.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java (hunk 4 above)
Comment (devpedrotavares) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java]: Done!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java (hunk 5 above)
Comment (devpedrotavares) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java]: Done!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java (hunk 6 above)
Comment (devpedrotavares) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java]: Done!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java (hunk 7 above)
Comment (devpedrotavares) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/test/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorTest.java]: Done!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java (hunk 8 above)
Comment (devpedrotavares) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java]: Done!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java (hunk 9 above)
Comment (devpedrotavares) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java]: Alright. Done!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java (hunk 10 above)
Comment (devpedrotavares) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java:169]: I made it case-insensitive, because there is already a validation being done in the front end (although it still allows the submission of the structure) and it's case-insensitive. 

Also, regarding the field names, I noticed that, in addition to being validated in `_validateDDMFormFieldNames`, they are being validated again in `_validateDDMFormFields`. This second one, however, does some extra checks and it is case-insensitive. 
I wonder if this should be refactored in order to be done in just one place.

**Comment by devpedrotavares** [comment]
ci:test:sf

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java (hunk 10 above)
Comment (carolmariaabb) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java:169]: Got it, @devpedrotavares,

> I wonder if this should be refactored in order to be done in just one place.
No need for now.

One more question, will this validation work if we have two fields: "FieldReference" and "fieldReference"? 
The exception should be thrown in this case, right?

Comment (devpedrotavares) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java:169]: Yes, the exception would be thrown in this case and the structure wouldn't be created.

Comment (carolmariaabb) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java:169]: @devpedrotavares Can you change the test to test this case as well? Instead of `Reference1` you can change to `fieldReference1` and `FieldReference1`? Thanks!

Comment (devpedrotavares) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-validator/src/main/java/com/liferay/dynamic/data/mapping/validator/internal/DDMFormValidatorImpl.java:169]: Done! I also changed the `Reference`'s to `FieldReference`s in `testDuplicateFieldName` for consistency.

**Comment by devpedrotavares** [comment]
ci:test:sf

**Comment by carolmariaabb** [comment]
ci:test:relevant

**Review by carolmariaabb — APPROVED** [review]
Great job, @devpedrotavares!

**Comment by carolmariaabb** [comment]
ci:reevaluate:1327401_7404

**Comment by carolmariaabb** [comment]
@devpedrotavares could you please check if the following tests are failing because of your changes?
`DataRecordResourceTest`, `DataDefinitionResourceTest` and  `DataLayoutResourceTest`.

Thanks!

**Comment by carolmariaabb** [comment]
@devpedrotavares please also fix the semantic version by running `gw baseline` in `/apps/data-engine/data-engine-rest-api` and `/apps/dynamic-data-mapping/dynamic-data-mapping-api`

**Comment by devpedrotavares** [comment]
> @devpedrotavares could you please check if the following tests are failing because of your changes? `DataRecordResourceTest`, `DataDefinitionResourceTest` and `DataLayoutResourceTest`.
> 
> Thanks!

Yes they were failing because of my changes. I updated them and now they're passing.

**Comment by devpedrotavares** [comment]
> @devpedrotavares please also fix the semantic version by running `gw baseline` in `/apps/data-engine/data-engine-rest-api` and `/apps/dynamic-data-mapping/dynamic-data-mapping-api`

Done!

**Comment by devpedrotavares** [comment]
ci:test:sf

**Comment by devpedrotavares** [comment]
ci:test:relevant

**Comment by devpedrotavares** [comment]
ci:reevaluate:1367401_2667

**Comment by devpedrotavares** [comment]
ci:test:relevant

**Comment by devpedrotavares** [comment]
ci:test:relevant

**Comment by devpedrotavares** [comment]
ci:test:relevant

**Comment by marcelabc** [comment]
ci:reevaluate:1277401_6294

**Review by marcelabc — APPROVED** [review]
LGTM 🎉

**Comment by marcelabc** [comment]
ci:test:sf

**Comment by marcelabc** [comment]
ci:test:relevant

**Comment by marcelabc** [comment]
ci:reevaluate:137null_7510

**Comment by marcelabc** [comment]
ci:test:relevant

**Comment by marcelabc** [comment]
ci:forward
